src/binary_annotator.cpp

```diff
@@ -0,0 +1,768 @@
+  uint64_t i = field_offset_start;
+  while (region_index < regions.size() && i < table_end_offset) {
+    const uint64_t region_start = regions[region_index].offset;
+    const uint64_t region_end = region_start + regions[region_index].length;
+
+    if (i < region_start) {
+      const uint64_t pad_bytes = region_start - i;
+      // We are at an index that is lower than any region, so pad upto its
+      // offset.
+      padding_regions.emplace_back(
+          MakeBinaryRegion(i, pad_bytes * sizeof(uint8_t),
+                           BinaryRegionType::Uint8, pad_bytes, 0, "padding"));
+      i = region_end;
+      region_index++;
+    } else if (i < region_end) {
+      i = region_end + 1;
+    } else {
+      region_index++;
+    }
+  }
+
+  // Handle the case where there is padding after the last known binary
+  // region. Calculate where we left off towards the expected end of the
+  // table.
+  if (i < table_end_offset) {
+    const uint64_t pad_bytes = table_end_offset - i + 1;
+    padding_regions.emplace_back(
+        MakeBinaryRegion(i - 1, pad_bytes * sizeof(uint8_t),
+                         BinaryRegionType::Uint8, pad_bytes, 0, "padding"));
+  }
+
+  regions.insert(regions.end(), padding_regions.begin(), padding_regions.end());
+
+  std::stable_sort(regions.begin(), regions.end(),
+                   [&](const BinaryRegion &a, const BinaryRegion &b) {
+                     return a.offset < b.offset;
+                   });
+
+  sections_.insert(std::make_pair(
+      table_offset,
+      MakeBinarySection(table->name()->str(), type, std::move(regions))));
+}
+
+uint64_t BinaryAnnotator::BuildStruct(uint64_t offset,
+                                      std::vector<BinaryRegion> &regions,
+                                      const reflection::Object *object) {
+  if (!object->is_struct()) { return offset; }
+
+  // Loop over all the fields in increasing order
+  ForAllFields(object, /*reverse=*/false, [&](const reflection::Field *field) {
+    if (IsScalar(field->type()->base_type())) {
+      offset = BuildStructureField(offset, object, field, regions);
+
+    } else if (field->type()->base_type() == reflection::BaseType::Obj) {
+      // Structs are stored inline, even when nested.
+      offset = BuildStruct(offset, regions,
+                           schema_->objects()->Get(field->type()->index()));
+    } else if (field->type()->base_type() == reflection::BaseType::Array) {
+      // Arrays are just repeated structures.
+      if (IsScalar(field->type()->element())) {
+        offset = BuildArrayField(offset, object, field,
+                                 field->type()->fixed_length(), regions);
+      } else {
+        for (uint16_t i = 0; i < field->type()->fixed_length(); ++i) {
+          // TODO(dbaileychess): This works, but the comments on the fields lose
+          // some context. Need to figure a way how to plumb the nested arrays
+          // comments together that isn't too confusing.
```

Comment:
Idea: Maybe you could plumb things through, xpath style?

`.struct1_field.struct2_array_field[0].struct3_field` this loses the type information of struct1, struct2, and struct3 though.

(no need to actually do this of course, but if you like it, consider updating the TODO)

Reply:
Ack.